core/src/main/java/hudson/PluginWrapper.java

```diff
@@ -377,6 +380,31 @@ public boolean isDeprecated() {
         return !getDeprecations().isEmpty();
     }
 
+    /**
+     * Inject the specified jar file(s) to the plugins classpath. 
+     * <strong>Warning:</strong> This is advanced usage that you should not need in
+     * 99.9% of all cases, and any jar insertion should happen early into the plugins lifecycle to prevent classloading
```

Comment:
is there a companion fix we could do so that you do not have to extend the `Plugin` class to use this in a plugin?

Reply:
so the Init  Milestones are a bit messy, ideally we could use those, but it is not a simple patch IIUC and would have a large potential to break things. (currently Initializers at early stages are not actually run at those stage - but either later or never (I do not recall which) - but if we supported that it could break some plugins that had some bad behaviour - especially if they expected a well formed classpath when they ran.

https://github.com/jenkinsci/jenkins/blob/ad427ecd669b243a5f6acddbe1497915ac171fee/core/src/main/java/hudson/PluginManager.java#L587 sorts out the dependencies - but strangely does not need to complete until Jenkins is fully up.

Extensions are only available after `EXTENSIONS_AUGMENTED` is attained (so initialized during that phase) which is after `PLUGINS_STARTED` so any `Plugin` (`* extends Plugin`) that depends on BC and has some static code to initialize etc would potentially not see the correct classpath anymore.

I'm happy for suggestions if we can avoid this new API and do something else instead, perhaps @jglick is more knowledable in this area may have some pointers?